src/duration/mod.rs

```diff
@@ -0,0 +1,356 @@
+//! This module contains a common [`Duration`] struct which is able to parse
+//! human-readable duration formats, like `2y 2h 20m 42s` or`15d 2m 2s`. It
+//! additionally implements many required traits, like [`Derivative`],
+//! [`JsonSchema`], [`Deserialize`][serde::Deserialize], and
+//! [`Serialize`][serde::Serialize].
+//!
+//! Furthermore, it implements [`Deref`], which enables us to use all associated
+//! functions of [`std::time::Duration`] without re-implementing the public
+//! functions on our own type.
+//!
+//! All operators should opt for [`Duration`] instead of the plain
+//! [`std::time::Duration`] when dealing with durations of any form, like
+//! timeouts or retries.
+
+use std::{
+    fmt::{Display, Write},
+    num::ParseIntError,
+    ops::{Deref, DerefMut},
+    str::FromStr,
+};
+
+use derivative::Derivative;
+use schemars::JsonSchema;
+use strum::IntoEnumIterator;
+use thiserror::Error;
+
+mod serde_impl;
+pub use serde_impl::*;
+
+#[derive(Debug, Error, PartialEq)]
+pub enum DurationParseError {
+    #[error("invalid input, either empty or contains non-ascii characters")]
+    InvalidInput,
+
+    #[error("failed to parse duration fragment")]
+    FragmentError(#[from] DurationFragmentParseError),
+}
+
+#[derive(Clone, Copy, Debug, Derivative, Hash, PartialEq, PartialOrd, JsonSchema)]
+pub struct Duration(std::time::Duration);
+
+impl FromStr for Duration {
+    type Err = DurationParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let input = s.trim();
+
+        // An empty or non-ascii input is invalid
+        if input.is_empty() || !input.is_ascii() {
+            return Err(DurationParseError::InvalidInput);
+        }
+
+        // Let's split up individual parts separated by a space
+        let parts: Vec<&str> = input.split(' ').collect();
+
+        // Parse each part as a DurationFragment and extract the final duration
+        // of each fragment in milliseconds
+        let values: Vec<u128> = parts
+            .iter()
+            .map(|p| p.parse::<DurationFragment>())
+            .map(|r| r.map(|f| f.millis()))
+            .collect::<Result<Vec<_>, DurationFragmentParseError>>()?;
+
+        // NOTE (Techassi): This derefernce is super weird, but
+        // Duration::from_millis doesn't accept a u128, but returns a u128
+        // when as_millis is called.
+        Ok(Self(std::time::Duration::from_millis(
+            values.iter().fold(0, |acc, v| acc + (*v as u64)),
+        )))
+    }
+}
+
+impl Display for Duration {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        // If the inner Duration is zero, print out '0ms' as milliseconds
+        // is the base unit for our Duration.
```

Comment:
Seconds are the SI base unit of time, milliseconds are subdivisions of that. (Following the logic of using the smallest known unit, would we want to switch to `0ns` if we added support for that? I'd say no.

Reply:
Changed in 9e9257d

Reply:
I'm fine with supporting ms if the user explicitly puts them in, I just think `0s` makes more sense as the fallback.

Reply:
Okay. Then I will re-add `ms` support, but default to `0s` for the `is_zero` case.

Reply:
See 58b04c3